### PR TITLE
TIP-1240: Add doc to add new attribute types

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
+# Add new attribute type
+
+## Context
+
+This documentation will help you to add a new attribute type in the PIM. 
+We will add a new attribute type called `Range`, defined with a min and a max value.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: defined by a min and a max value

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+
+**Warnings**
+
+- It does not imply the search on the values defined on this attribute
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: It does not cover [***]

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+- It does not imply the indexation of this field
+- It does not imply the API import nor export processes
+- It does not do any validation on the Range values (min inferior to max, etc)
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: It does not perform

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('services.yml');
+    }
+}
+```
+
+## Step 2: Create the attribute type (backend)
+
+Attribute types are auto loaded with `RegisterAttributeTypePass`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: automatically

Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: with > by the

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('services.yml');
+    }
+}
+```
+
+## Step 2: Create the attribute type (backend)
+
+Attribute types are auto loaded with `RegisterAttributeTypePass`. 
+You need to create a new `AttributeTypeInterface` and add its service. For it, you can use the `AbstractAttributeType`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: and add its definition as a service

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+        tags:
+            - { name: pim_catalog.attribute_type, alias: range, entity: '%pim_catalog.entity.product.class%' }
+```
+
+## Step 3: Create the values
+
+**Warning** For now, you have to create 2 value factories (one for reading purpose, one for writing purpose).
+They are very similar, and will probably be refactored in the next months.
+
+To keep this tutorial as the simplest possible, we will use the same format, from back-end to front-end.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: For the sake of simplicity, we will re-use the same format both for back-end and front-end

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+
+    public function supportedAttributeType(): string
+    {
+        return RangeType::RANGE;
+    }
+}
+```
+
+```yaml
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: why did you add [...] ad the end of this comment?

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        arguments:
+            - 'Akeneo\Pim\Enrichment\Component\Product\Value\ScalarValue'
+            - 'range'
+        tags:
+            - { name: pim_catalog.factory.value }
+```
+
+## Step 4: Process the values
+
+You will need 2 new services to process the values.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: values: a setter and a comparator.

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+First, the setter, used on the product updater, to update the entity values.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: The setter, used by the product updater, responsible for updating entity values

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+For this one, as there is no specific treatment, we will use the default `AttributeSetter`:
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: processing instead of treatment

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
+    acme.range.updater.setter.range_value:
+        class: '%pim_catalog.updater.setter.value.class%'
+        parent: pim_catalog.updater.setter.abstract
+        arguments: [['range']]
+        tags:
+            - { name: 'pim_catalog.updater.setter' }
+```
+
+Next, the comparator, used to know if the value was updated or not. A comparator returns null if the data was not updated.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: if > whether

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+    acme.range.comparator.attribute.range:
+        class: Acme\RangeBundle\Product\Comparator\Attribute\RangeComparator
+        tags:
+            - { name: pim_catalog.attribute.comparator, priority: -100 }
+```
+
+## Step 5: Completeness
+
+To compute the completeness, you will define when a value is **complete** or **incomplete**.
+Since 4.0, we use the notion of masks to generate keys for each filled value.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Indicate which software you are referring to here

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+The generated mask for a filled value should looks like `attributeCode-channelCode-localeCode`.
+You need to define your own mask generator for your new attribute type.
+In our case, we will define a value as **complete** only of the `min` and `max` are filled.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: if

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+```yaml
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
+    acme.range.completeness.mask_item_generator.range:
+        class: Acme\RangeBundle\Product\Completeness\MaskItemGenerator\RangeMaskItemGenerator
+        tags: [{ name: akeneo.pim.enrichment.completeness.mask_item_generator }]
+```
+
+## Step 6: User Interface
+
+You will need to display a new field in the PEF to manipulate your new attribute type.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Product Edit Form

Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: To be able to use your new attribute type

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+Simply add a new HTML template with 2 fields:
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Remove "Simply"

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+                    min: this.$('.min').val(),
+                    max: this.$('.max').val()
+                });
+            }
+        });
+    }
+);
+```
+
+Register it in the requirejs file:
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Add it as a new entry in your `requirejs.yml` file

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+```yaml
+# src/Acme/RangeBundle/Resources/config/requirejs.yml
+config:
+    paths:
+        acme/range-field: acmerange/js/product/field/range-field
+        acme/template/range-field: acmerange/templates/product/field/range-field.html
+```
+
+Next, you will need to tell the front what module to use regarding your new attribute type.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: indicate

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+The attribute types it can be manipulated through UI have to own a `FieldProviderInterface`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Unclear what you meant here

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 19]
```diff
@@ -0,0 +1,476 @@
...
+Define this new provider, it supports only `range` attribute types.
+This class will send to front-end the JS extension to use (here, `acme-range-field`)
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: will indicate what JS extension to use

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 20]
```diff
@@ -0,0 +1,476 @@
...
+    &--range {
+      background-image: url("/bundles/pimui/images/attribute/icon-metric.svg");
+    }
+  }
+}
+```
+
+## Step 7: Export values
+
+If you want to be able to export values through flat files, you need to code how to convert a range value into a cell.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: at the beginning of the tutorial you mentioned the export was not covered, am I missing something?

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 21]
```diff
@@ -0,0 +1,476 @@
...
+We will add a new attribute type called `Range`, defined with a min and a max value.
+
+**Warnings**
+
+- It does not imply the search on the values defined on this attribute
+- It does not imply the indexation of this field
+- It does not imply the API import nor export processes
+- It does not do any validation on the Range values (min inferior to max, etc)
+
+## Step 1: Initialize your Symfony bundle
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: I'm not sure about this part. IMO if people want to add a new attribute type, they already know how to create a bundle, register services, etc.

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 22]
```diff
@@ -0,0 +1,476 @@
...
+<?php #src/Acme/RangeBundle/AcmeRangeBundle.php
+
+namespace Acme\RangeBundle;
+
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class AcmeRangeBundle extends Bundle {}
+```
+
+Temporary, create an empty service file
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
Create a (temporarily) empty service file
```

Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Create a new service file that will be used later.

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 4 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: I'm not sure the sentence actually brings value here, while it helps understanding what happens behind the scenes, it is not necessary for global comprehension. Or you can move this to the sentence just below like this: 
> Add this class in services.yml with the `pim_catalog.attribute_type` tag. Please note that services with such a tag will be loaded by the `RegisterAttributePass`.

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 23]
```diff
@@ -0,0 +1,476 @@
...
+use Acme\RangeBundle\AttributeType\RangeType;
+use Acme\RangeBundle\Product\Value\RangeValue;
+use Akeneo\Pim\Enrichment\Component\Product\Factory\Read\Value\ValueFactory;
+use Akeneo\Pim\Enrichment\Component\Product\Model\ValueInterface;
+use Akeneo\Pim\Structure\Component\Query\PublicApi\AttributeType\Attribute;
+use Akeneo\Tool\Component\StorageUtils\Exception\InvalidPropertyTypeException;
+
+class RangeValueFactory implements ValueFactory
+{
+    function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Please mind line length

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 24]
```diff
@@ -0,0 +1,476 @@
...
+    }
+}
+```
+
+```yaml
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
+    acme.range.comparator.attribute.range:
+        class: Acme\RangeBundle\Product\Comparator\Attribute\RangeComparator
+        tags:
+            - { name: pim_catalog.attribute.comparator, priority: -100 }
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Why such priority number?

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 25]
```diff
@@ -0,0 +1,476 @@
...
+```
+
+## Step 5: Completeness
+
+To compute the completeness, you will define when a value is **complete** or **incomplete**.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: > To be able to compute the completeness, it has to be defined when a value is considered to be **complete** or **incomplete**

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 26]
```diff
@@ -0,0 +1,476 @@
+# Add new attribute type
+
+## Context
+
+This documentation will help you to add a new attribute type in the PIM. 
+We will add a new attribute type called `Range`, defined with a min and a max value.
+
+**Warnings**
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: For all this part: IMO it makes your new AttributeType almost useless. In this paragraph I would rather explain:
- whet id _does_ cover
- for what it does not (filters, ...), at least point out where one should look at in order to take example, e.g:
``` 
if you want to go further, have a look at these services:
- ...
- ...
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 27]
```diff
@@ -0,0 +1,476 @@
...
+        class: Acme\RangeBundle\Product\Comparator\Attribute\RangeComparator
+        tags:
+            - { name: pim_catalog.attribute.comparator, priority: -100 }
+```
+
+## Step 5: Completeness
+
+To compute the completeness, you will define when a value is **complete** or **incomplete**.
+Since 4.0, we use the notion of masks to generate keys for each filled value.
+The generated mask for a filled value should looks like `attributeCode-channelCode-localeCode`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: Can you tell us more here about the usage of masks, why their uniqueness is important, maybe show a sample from the database?

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 28]
```diff
@@ -0,0 +1,476 @@
...
+{
+    protected function prepareData(AttributeInterface $attribute, $data, bool $ignoreUnknownData)
+    {
+        if ($data === null) {
+            $data = ['min' => null, 'max' => null];
+        }
+        if (!is_array($data)) {
+            throw InvalidPropertyTypeException::arrayExpected($attribute->getCode(), static::class, $data);
+        }
+        if (!key_exists('min', $data)) {
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: please use `array_key_exists` (most commonly used form, `key_exists` is an alias)

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 29]
```diff
@@ -0,0 +1,476 @@
...
+            throw InvalidPropertyTypeException::arrayKeyExpected($attribute->getCode(), 'min', static::class, $data);
+        }
+        if (!key_exists('max', $data)) {
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: same

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 23 above)
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 30]
```diff
@@ -0,0 +1,476 @@
...
+
+## Step 6: User Interface
+
+You will need to display a new field in the PEF to manipulate your new attribute type.
+Simply add a new HTML template with 2 fields:
+
+```html
+<!-- src/Acme/RangeBundle/Resources/public/templates/product/field/range-field.html -->
+<input class="AknTextField AknTextField--noRightRadius min" value="<%- value.data.min %>">
+<input class="AknTextField AknTextField--noLeftRadius max" value="<%- value.data.max %>">
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: You can add a link here to the style guide in case someone wants to create a Christmas tree as a field hehe

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 31]
```diff
@@ -0,0 +1,476 @@
...
+Register it in the requirejs file:
+```yaml
+# src/Acme/RangeBundle/Resources/config/requirejs.yml
+config:
+    paths:
+        acme/template/range-field: acmerange/templates/product/field/range-field.html
+```
+
+Next, you have to create the linked JS module. 
+The `updateModel` function will be called when the user will write in one of the inputs.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: each time users will fill in the fields

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 32]
```diff
@@ -0,0 +1,476 @@
...
+      background-image: url("/bundles/pimui/images/attribute/icon-metric.svg");
+    }
+  }
+}
+```
+
+## Step 7: Export values
+
+If you want to be able to export values through flat files, you need to code how to convert a range value into a cell.
+Here, we render a 10 to 20 range value in this format: `[10...20]`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: The value will be normalized like this: [10...20]

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 33]
```diff
@@ -0,0 +1,476 @@
...
+{
+    function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+    {
+        if ($data === null) {
+            $data = ['min' => null, 'max' => null];
+        }
+        if (!is_array($data)) {
+            throw InvalidPropertyTypeException::arrayExpected($attribute->code(), static::class, $data);
+        }
+        if (!key_exists('min', $data)) {
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
        if (!array_key_exists('min', $data)) {
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 34]
```diff
@@ -0,0 +1,476 @@
+# Add new attribute type
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: > Add **a** new attribute type 

or 

> Add new attribute type**s**

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 26 above)
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: **Limitations**

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 27 above)
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
The generated mask for a filled value should look like `attributeCode-channelCode-localeCode`.
```

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 2 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: Can you point to some examples, namespaces to check, etc?

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 35]
```diff
@@ -0,0 +1,476 @@
...
+
+    public function getName()
+    {
+        return self::RANGE;
+    }
+}
+```
+
+Call this class in services file with the `pim_catalog.attribute_type` tag.
+You can customize the backend type to do special treatments on the validation and normalization processes. We will use the default one `text`.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: special processing for validation and for normalization.

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 36]
```diff
@@ -0,0 +1,476 @@
...
+    acme.range.attributetype.range:
+        class: Acme\RangeBundle\AttributeType\RangeType
+        arguments: ['text']
+        tags:
+            - { name: pim_catalog.attribute_type, alias: range, entity: '%pim_catalog.entity.product.class%' }
+```
+
+## Step 3: Create the values
+
+**Warning** For now, you have to create 2 value factories (one for reading purpose, one for writing purpose).
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: another one for writing purpose

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 37]
```diff
@@ -0,0 +1,476 @@
...
+They are very similar, and will probably be refactored in the next months.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: you can err on the side of caution here and mention a vague future ;)

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 17 above)
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: I did not understand this paragraph, maybe you should rephrase :man_shrugging:

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 34 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: Don't forget the title of this PR haha

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 38]
```diff
@@ -0,0 +1,476 @@
...
+        }
+        if (!is_array($data)) {
+            throw InvalidPropertyTypeException::arrayExpected($attribute->getCode(), static::class, $data);
+        }
+        if (!key_exists('min', $data)) {
+            throw InvalidPropertyTypeException::arrayKeyExpected($attribute->getCode(), 'min', static::class, $data);
+        }
+        if (!key_exists('max', $data)) {
+            throw InvalidPropertyTypeException::arrayKeyExpected($attribute->getCode(), 'max', static::class, $data);
+        }
```
Comment (mmetayer) [inline internal_doc/enrichment/add_new_attribute_type.md]: also: now you shouldn't be able to create a value where both `min` and `max` are null

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 37 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: We are currently working on it :)

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 38 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: You should for the PEF.

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 7 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: To explain the reader that there is still other services above... Don't know if it's a good idea.

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 20 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: The API export only :)

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 7 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: Oh I see. Personally I would add a new empty line and then
`# ...`

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 20 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: Gotcha

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 37 above)
Comment (rybus) [reply internal_doc/enrichment/add_new_attribute_type.md]: Awesome!

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 39]
```diff
@@ -0,0 +1,444 @@
...
+
+```yaml
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
+    acme.range.provider.field.range:
+        class: Acme\RangeBundle\Enrich\Provider\Field\RangeFieldProvider
+        tags:
+            - { name: pim_enrich.provider.field }
+```
+
+Then, register in the form extensions, that the `acme-range-field` have to load the `acme/range-field` extension defined above.
```
Comment (rybus) [inline internal_doc/enrichment/add_new_attribute_type.md]: has

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 40]
```diff
@@ -0,0 +1,444 @@
...
+
+This documentation will not:
+- cover the indexation of this field (you can take a look at `Akeneo\Pim\Enrichment\Component\Product\Normalizer\Indexing\Value\AbstractProductValueNormalizer`),
+- cover the search on the values defined on this attribute (you can take a look at `Akeneo\Pim\Enrichment\Component\Product\Query\Filter\AttributeFilterInterface`),
+- cover the import processes (take a look at `Akeneo\Pim\Enrichment\Component\Product\Connector\ArrayConverter\FlatToStandard\ValueConverter\ValueConverterInterface`).
+
+## Step 1: Create the attribute type (backend)
+
+You need to create a new `AttributeTypeInterface` and add its service. 
+For it, you can use the `AbstractAttributeType` and add its definition as a service.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
To do so, you can use the `AbstractAttributeType` and add its definition as a service.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 41]
```diff
@@ -0,0 +1,444 @@
...
+    public function getName()
+    {
+        return self::RANGE;
+    }
+}
+```
+
+Add this class in `services.yml` with the `pim_catalog.attribute_type` tag.
+Please note that services with such a tag will be loaded by the `RegisterAttributePass`.
+You can customize the backend type to do special processes for validation and normalization processes. We will use the default one `text`.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
You can customize the backend type for special validation and normalization processes. We will use the default one `text`. 
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 42]
```diff
@@ -0,0 +1,444 @@
...
+    acme.range.attributetype.range:
+        class: Acme\RangeBundle\AttributeType\RangeType
+        arguments: ['text']
+        tags:
+            - { name: pim_catalog.attribute_type, alias: range, entity: '%pim_catalog.entity.product.class%' }
+```
+
+## Step 2: Create the values
+
+**Warning** For now, you have to create 2 value factories (one for reading purpose, another one for writing purpose).
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
**Warning** For now, you have to create 2 value factories (one for reading purposes, another one for writing purposes).
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 43]
```diff
@@ -0,0 +1,444 @@
...
+# src/Acme/RangeBundle/Resources/config/services.yml [...]
+    acme.range.comparator.attribute.range:
+        class: Acme\RangeBundle\Product\Comparator\Attribute\RangeComparator
+        tags:
+            - { name: pim_catalog.attribute.comparator }
+```
+
+## Step 4: Completeness
+
+To be able to compute the completeness, it has to be defined when a value is considered to be **complete** or **incomplete**.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
To be able to compute the completeness, a value has to be defined as **complete** or **incomplete**.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 44]
```diff
@@ -0,0 +1,444 @@
...
+Since Akeneo PIM 4.0, we use the notion of masks to generate keys for each filled value.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
Since Akeneo PIM 4.0, we have been using the notion of masks to generate keys for each filled value.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 45]
```diff
@@ -0,0 +1,444 @@
...
+The generated mask for a filled value should look like `attributeCode-channelCode-localeCode`.
+For more information about completeness masks, please read documentation of `RequiredAttributesMask` class.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
For more information about completeness masks, please read the documentation of `RequiredAttributesMask` class.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 46]
```diff
@@ -0,0 +1,444 @@
...
+Add a new HTML template with 2 fields. 
+You can customize the style of your page using our [styleguide](https://docs.akeneo.com/master/design_pim/styleguide/index.php). 
+
+```html
+<!-- src/Acme/RangeBundle/Resources/public/templates/product/field/range-field.html -->
+<input class="AknTextField AknTextField--noRightRadius min" value="<%- value.data.min %>">
+<input class="AknTextField AknTextField--noLeftRadius max" value="<%- value.data.max %>">
+```
+
+Register it in the requirejs file:
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
Register it in the required file:
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 47]
```diff
@@ -0,0 +1,444 @@
...
+```yaml
+# src/Acme/RangeBundle/Resources/config/requirejs.yml
+config:
+    paths:
+        acme/template/range-field: acmerange/templates/product/field/range-field.html
+```
+
+Each attribute type (text, boolean, single option, etc) have a dedicated javascript module loaded in the product edit form.
+You have to create a new `range` javascript module to edit these values. 
+The `updateModel` function will be called each time the user will fill in the fields.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
The `updateModel` function will be called each time the user fills in the fields.  
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 48]
```diff
@@ -0,0 +1,444 @@
...
+```yaml
+# src/Acme/RangeBundle/Resources/config/requirejs.yml
+config:
+    paths:
+        acme/range-field: acmerange/js/product/field/range-field
+        acme/template/range-field: acmerange/templates/product/field/range-field.html
+```
+
+Next, the backend needs to indicate to the front what module to use regarding the attribute type.
+For that, each attribute type have a corresponding `FieldProviderInterface`.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
To that end, each attribute type has a corresponding `FieldProviderInterface`.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 49]
```diff
@@ -0,0 +1,444 @@
...
+Define this new provider, it supports only `range` attribute types.
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
Define this new provider, it only supports `range` attribute types.
```

File: internal_doc/enrichment/add_new_attribute_type.md [hunk 50]
```diff
@@ -0,0 +1,444 @@
...
+This class indicate what javascript extension to use (here, `acme-range-field`).
```
Comment (camillebourjade) [inline internal_doc/enrichment/add_new_attribute_type.md]:
```suggestion
This class indicates what javascript extension to use (here, `acme-range-field`).
```

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 44 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: It's the present/future :)

File: internal_doc/enrichment/add_new_attribute_type.md (hunk 46 above)
Comment (pierallard) [reply internal_doc/enrichment/add_new_attribute_type.md]: It's "requirejs", it's not a typo :)
